### DISCUSSION TITLE
Help with tmap in Interlude V

### DISCUSSION THREAD

**Discussion by brrgrl** [themetaschemer/malt#8] [Q&A] [answered]
Up to Interlude V, I have been getting by with just having `(require malt)` provide all the tools that I need while building along with the book. However, when I got to using `tmap` it was undefined.

My first few passes through the repo, it looks like it's not defined in flat-tensors but it is defined in nested-tensors.

So I'm a little lost. 

Defining `tmap` with `(define tmap vector-map)` and then trying the example leads to
```
> (tmap add1 (tensor 3 4 5))
. . add1: contract violation
  expected: number?
  given: #<procedure:flat>
> 
```

Instead requiring nested tensors `(require malt/nested-tensors)` I get something that works
```
> (require malt/nested-tensors)
> (tmap add1 (tensor 3 4 5))
#(4 5 6)
> 
```

**Comment by themetaschemer** [comment, marked as answer]
Hi Brrgrl,
For Interlude V, you have to do:

`(require malt/interlude-V)`

This is also mentioned in the chapter guide on the website, but perhaps we have been remiss in making it more prominent. 
The default implementation of flat-tensors implements ext1 and ext2 differently, so we have provided a separate module to allow you to experiment with the code in Interlude V.

**Reply by brrgrl** [reply]
Thank you!!

**Reply by brrgrl** [reply]
Follow up: when going through the rest of interlude-V, it looks like `flatten-2` should be provided but in no-overrides.rkt it isn't. 
Is this the case?

**Reply by themetaschemer** [reply]
Hmm ... that might be a bug. Let me investigate.

**Reply by themetaschemer** [reply]
Okay, I have fixed and pushed. Thank you for pointing it out.

**Reply by brrgrl** [reply]
Thank you!

I found a dot-product defined in G-correlate but I'm not sure if that's the one that is referenced by frame 43 in interlude-V.

Should the dot-product from malted with the `d-sum` substitution be included?
```
(define d-dot-product
  (λ (w t)
   (d-sum
     (* w t))))
```

**Reply by themetaschemer** [reply]
Brrgrl, sorry you're having so many issues with this. Yes, that definition will work. But I am also updating interlude-V to include a dot-product-1-1 explicitly. Please pull the latest.
